crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,342 @@ impl MeshletMesh {
-        }
-        if mesh.attributes().map(|(id, _)| id).ne([
-            Mesh::ATTRIBUTE_POSITION.id,
-            Mesh::ATTRIBUTE_NORMAL.id,
-            Mesh::ATTRIBUTE_UV_0.id,
-            Mesh::ATTRIBUTE_TANGENT.id,
-        ]) {
-            return Err(MeshToMeshletMeshConversionError::WrongMeshVertexAttributes);
-        }
-        let indices = match mesh.indices() {
-            Some(Indices::U32(indices)) => Cow::Borrowed(indices.as_slice()),
-            Some(Indices::U16(indices)) => indices.iter().map(|i| *i as u32).collect(),
-            _ => return Err(MeshToMeshletMeshConversionError::MeshMissingIndices),
-        };
-        let vertex_buffer = mesh.get_vertex_buffer_data();
-        let vertices =
-            VertexDataAdapter::new(&vertex_buffer, mesh.get_vertex_size() as usize, 0).unwrap();
-
-        // Split the mesh into meshlets
-        let meshopt_meshlets = build_meshlets(&indices, &vertices, 64, 64, 0.0);
+        let indices = validate_input_mesh(mesh)?;
 
-        // Calculate meshlet bounding spheres
-        let meshlet_bounding_spheres = meshopt_meshlets
+        // Split the mesh into an initial list of meshlets (LOD 0)
+        let vertex_buffer = mesh.get_vertex_buffer_data();
+        let vertex_stride = mesh.get_vertex_size() as usize;
+        let vertices = VertexDataAdapter::new(&vertex_buffer, vertex_stride, 0).unwrap();
+        let mut meshlets = compute_meshlets(&indices, &vertices);
+        let mut bounding_spheres = meshlets
             .iter()
-            .map(|meshlet| {
-                compute_meshlet_bounds_decoder(
-                    meshlet,
-                    mesh.attribute(Mesh::ATTRIBUTE_POSITION)
-                        .unwrap()
-                        .as_float3()
-                        .unwrap(),
-                )
-            })
-            .map(|bounds| MeshletBoundingSphere {
-                center: bounds.center.into(),
-                radius: bounds.radius,
+            .map(|meshlet| compute_meshlet_bounds(meshlet, &vertices))
+            .map(convert_meshlet_bounds)
+            .map(|bounding_sphere| MeshletBoundingSpheres {
+                self_culling: bounding_sphere,
+                self_lod: MeshletBoundingSphere {
+                    center: bounding_sphere.center,
+                    radius: 0.0,
+                },
+                parent_lod: MeshletBoundingSphere {
+                    center: bounding_sphere.center,
+                    radius: f32::MAX,
+                },
             })
-            .collect();
+            .collect::<Vec<_>>();
+        let worst_case_meshlet_triangles = meshlets
+            .meshlets
+            .iter()
+            .map(|m| m.triangle_count as u64)
+            .sum();
+
+        // Build further LODs
+        let mut simplification_queue = 0..meshlets.len();
+        let mut lod_level = 1;
+        while simplification_queue.len() > 1 {
```

Comment:
This feels very much like it could be a for loop, which would feel less error prone.

Reply:
It's not a for loop. We dynamically adjust the queue range to process more items at the end of the loop.

Iter 1: 0..1000
Iter 2: 1000..1500
Iter 3: 1500..1700
...

Reply:
Ahhh, meshlets is mutated in split_simplified_groups_into_new_meshlets. I see.